### DISCUSSION TITLE
Can we use multiple second-level domains with multitenancy?

### DISCUSSION THREAD

**Discussion by iabduul7** [spatie/laravel-multitenancy#372] [Q&A] [answered]
I have implemented the simplest example using the [Spatie docs for multitenancy](https://spatie.be/docs/laravel-multitenancy/v2/introduction), that is working perfectly fine. Now, I intend to use multiple second-level domains for each tenant I have. 

For example; I have 2 tenants `company-a` and `company-b` and they are being served at `company-a.localhost` and `company-b.localhost`, now what I want is that when I visit `company-a.admin.localhost`, it should tell me **COMPANY-A ADMIN** and If I visit `company-a.employee.localhost`, it should tell me **COMPANY-A EMPLOYEE**. 

I have tried using `subdomain` on routes in `RouteServiceProvider` like the following:

          Route::middleware('web')
                ->group(base_path('routes/security.php'));

           Route::domain($this->baseDomain('admin'))
                ->middleware('web')
                ->name('admin.')
                ->group(base_path('routes/admin.php'));

            Route::domain($this->baseDomain('employee'))
                ->middleware('web')
                ->name('employee.')
                ->group(base_path('routes/employee.php'));

Without subdomain, it works fine, but the routes with second-level domain, it falls to base level domain route and does not get the current tenant.
What am I missing here? Is this even possible to implement. 

Thankyou.

**Comment by rsenses** [comment]
I've done something similar modifying DomainTenantFinder and without Route:domain.
The idea is to return the tenant in the landlord database no matter the strategy.
Take a look at https://spatie.be/docs/laravel-multitenancy/v2/basic-usage/automatically-determining-the-current-tenant/

**Comment by iabduul7** [comment, marked as answer]
I have achieved this by using some checks, in `RouteServiceProvider`, I have not used the actual `domain` function on `Route` like we do normally i.e. `Route::domain('foo.bar')`. The reason was that, the `Spatie` package use a kind of middleware `Spatie\Multitenancy\TenantFinder\DomainTenantFinder::class` which runs whenever we hit the domain with tenant `comapny-a.localhost`. And it gets the tenant from hostname i.e `comapny-a.localhost`.

```
    public function findForRequest(Request $request):?Tenant
    {
        $host = $request->getHost();
        return $this->getTenantModel()::whereDomain($host)->first();
    }
```

In my `RouteServiceProvide`:
```
     $this->routes(function () {
         $class = 'security';
         $middleware = 'web';
         if (Str::contains(request()->getHost(), 'admin')) {
             $class = 'admin';
         } elseif (Str::contains(request()->getHost(), 'employee')) {
             $class = 'employee';
         } elseif (Str::contains(request()->getHost(), 'api')) {
             $class = 'api';
             $middleware = 'api';
         }

         Route::middleware($middleware)
             ->name("$class.")
             ->group(base_path("routes/${class}.php"));
     });
```

As In my scenario, I had only these 2 kind of second-level domains and so, I just checked if this particular keyword exists in the hostname and choosing the file and middleware accordingly.

I also overrided the `DomainTenantFinder` class and in `multitenancy.php` config file: 

```
    public function findForRequest(Request $request): ?Tenant
    {
        $host = $request->getHost();
        $host = str_replace('admin.', '', $host);
        $host = str_replace('employee.', '', $host);
        $host = str_replace('api.', '', $host);
        $tenant = $this->getTenantModel()::whereDomain($host)->first();
        if (empty($tenant)) {
            abort(404);
        }
        return $tenant;
    }
```

I have acheived the desired outcome, however, I have a security concern, specially in `RouteServiceProvider` logic. Thoughts??

**Reply by rsenses** [reply]
I think a simple aproach would be to use the original DomainTenantFinder an change the domain column on landlord database to company-a.admin.localhost and company-a.employee.localhost

**Reply by iabduul7** [reply]
Well there are two things to consider here
- There are separate `route` files for **admin**, **employee**, **security** and **api**... As you have seen my `RouteServiceProvider`, so it should be kept that way. 

- Adding more columns like `domain` i.e. `admin_domain`, `employee_domain` to find and get `tenant`, would that be recommended way?


PS: The only think that is making me to think is whatever I have done in `RouteServiceProvider`, clearly the `domain` method doesn't seem to be working in this particular scenario as it gives 404.

**Reply by iabduul7** [reply]
I have modified my logic a bit. in `RouteServiceProvider`, the `baseDomain` function was returning `xyz.{{base_domain}}` but the `DomainFinder` was looking for `{{tenant}}.xyz.{{base_domain}}`, so in order to give the class this very response, I modified my `baseDomain` function and used the `domain` approach in my `Routes`.

```
  private function baseDomain(string $subdomain = ''): string
   {
        $baseDomain = config('app.base_domain');

        if (strlen($subdomain) > 0) {
            $subdomain = "{$subdomain}.";
        }

        $tenant = app('currentTenant');
        if (strlen($tenant) > 0) {
            $tenant = "{$tenant->name}.";
        }

        return "{$tenant}{$subdomain}{$baseDomain}";
    }
```

**Comment by MattWiseParking** [comment]
Even simple answer, just simply extending the DomainTenantFinder add the findForRequest method and add in one line.

```php
<?php

namespace App\Multitenancy;

use Illuminate\Http\Request;
use Spatie\Multitenancy\Models\Tenant;
use \Spatie\Multitenancy\TenantFinder\DomainTenantFinder as TenantFinder;

class DomainTenantFinder extends TenantFinder
{
    public function findForRequest(Request $request): ?Tenant
    {
        $host = $request->getHost();
        $host = (str_contains($host,'admin.'))?str_replace('admin.', '', $host):$host;

        return $this->getTenantModel()::whereDomain($host)->first();
    }
}
```

**Reply by MattWiseParking** [reply]
if you have multiple custom second level domains, use an array as a protected property, then search the array instead of str_contains

**Reply by MattWiseParking** [reply]
dont forget to change the multitenancy config file and point the tenant_finder to your new class

`'tenant_finder' => \App\Multitenancy\DomainTenantFinder::class,`
